### ISSUE TITLE
Touch & Stylus

### ISSUE THREAD

**Issue by vizv** [danielquinn/Gentoo-Surface-Pro-3#13]
After apply the `x86-input-evdev` patch, the other button on the stylus works however the touchscreen stops working. 

ArchWiki provides [an alternative way](https://wiki.archlinux.org/index.php/Microsoft_Surface_Pro_3#Tuning_the_Pen) to solve the same problem which was from an [reddit post](https://www.reddit.com/r/SurfaceLinux/comments/3mu28a/sp3_pen_tip_button_working/)

I tried to edit `10-evdev.conf` and `50-wacom.conf` but it crashes my X Server. After check the log file of X Server, I leave `10-evdev.conf` unmodified then both stylus buttons and touchscreen works now.

However I still can't get the bluetooth and tip button of stylus working (crashes after pairing with stylus).

**Comment by danielquinn** [comment]
Wow, you've done a lot more work in this area than I have (I never use the stylus, so to be honest, I hadn't noticed that the patch wasn't working).  Do you think you could roll your changes and adaptation of the reddit post into a pull request?  I understand that the tip button isn't working, but it sounds like you've got a better-working setup than others using this repo.

**Comment by joelacrisp** [comment]
FYI, I got the blue button working, see the instructions in my fork of the repo.

Unfortunately I can't run Gentoo on my Surface right now - doing too much Windows work. But keep up the good fight!

**Comment by vizv** [comment]
@danielquinn I'll create a PR later today. There will also be a WiFi fix for a bug introduced in higher kernel version (I'm in 4.5.3 right now), related information can be found in a [kernel bug tracker](https://bugzilla.kernel.org/show_bug.cgi?id=109681).

@joelacrisp Thanks, however I don't see your fork of this repo, I have googled around for hours about the bluetooth issue but found nothing, when I try to pair with the stylus, it gives a segfault in dmesg:

```
bluetoothd[2349]: segfault at 20 ip 0000555f56a849fc sp 00007ffce35f6370 error 4 in bluetoothd[555f56a3f000+10c000]
```

Here are some info for my setup:

```
net-wireless/bluez
Installed versions:  5.37(10:12:25 PM 11/05/16)(cups obex readline systemd udev -debug -doc -extra-tools -selinux -test -test-programs ABI_MIPS="-n32 -n64 -o32" ABI_PPC="-32 -64" ABI_S390="-32 -64" ABI_X86="32 64 -x32" PYTHON_TARGETS="python2_7")
```

Kernel configuration related to bluetooth:

```
# grep 'CONFIG_BT_' .config
CONFIG_BT_BREDR=y
CONFIG_BT_RFCOMM=y
# CONFIG_BT_RFCOMM_TTY is not set
CONFIG_BT_BNEP=y
# CONFIG_BT_BNEP_MC_FILTER is not set
# CONFIG_BT_BNEP_PROTO_FILTER is not set
CONFIG_BT_HIDP=y
CONFIG_BT_HS=y
CONFIG_BT_LE=y
# CONFIG_BT_6LOWPAN is not set
# CONFIG_BT_SELFTEST is not set
CONFIG_BT_DEBUGFS=y
CONFIG_BT_INTEL=y
CONFIG_BT_BCM=y
CONFIG_BT_RTL=y
CONFIG_BT_HCIBTUSB=y
CONFIG_BT_HCIBTUSB_BCM=y
CONFIG_BT_HCIBTUSB_RTL=y
# CONFIG_BT_HCIBTSDIO is not set
CONFIG_BT_HCIUART=y
CONFIG_BT_HCIUART_H4=y
# CONFIG_BT_HCIUART_BCSP is not set
# CONFIG_BT_HCIUART_ATH3K is not set
# CONFIG_BT_HCIUART_LL is not set
# CONFIG_BT_HCIUART_3WIRE is not set
CONFIG_BT_HCIUART_INTEL=y
CONFIG_BT_HCIUART_BCM=y
# CONFIG_BT_HCIUART_QCA is not set
# CONFIG_BT_HCIBCM203X is not set
# CONFIG_BT_HCIBPA10X is not set
# CONFIG_BT_HCIBFUSB is not set
# CONFIG_BT_HCIDTL1 is not set
# CONFIG_BT_HCIBT3C is not set
# CONFIG_BT_HCIBLUECARD is not set
# CONFIG_BT_HCIBTUART is not set
# CONFIG_BT_HCIVHCI is not set
CONFIG_BT_MRVL=y
# CONFIG_BT_MRVL_SDIO is not set
# CONFIG_BT_ATH3K is not set
```

**Comment by joelacrisp** [comment]
It is in my fork under my other GIT ID cydergoth
Ah! It was in the PR comments :-( the PR seems to be gone now.

**Comment by cydergoth** [comment]
https://github.com/danielquinn/Gentoo-Surface-Pro-3/pull/4

**Comment by vizv** [comment]
Since I'm using kernel 4.5.3, Could someone help me verify if the following patch works for lower kernel version? I have my SecureBoot configured without any bootloader, so replace a lower version kernel can be painful.

You'll need `INPUT_DEVICES="synaptics evdev wacom"` for your `make.conf` so x11-drivers/xf86-input-wacom will be pulled in.

And then apply the patch for `/usr/share/X11/xorg.conf.d/50-wacom.conf`

```diff
--- a/50-wacom.conf     2016-05-12 16:46:00.906184385 -0300
+++ b/50-wacom.conf     2016-05-11 21:43:31.741441494 -0300
@@ -51,7 +51,7 @@
 # N-Trig Duosense Electromagnetic Digitizer
 Section "InputClass"
        Identifier "Wacom N-Trig class"
-       MatchProduct "HID 1b96:0001|N-Trig Pen|N-Trig DuoSense"
+       MatchProduct "HID 1b96:0001|N-Trig Pen|N-Trig DuoSense|1B96:1B05 Pen"
        MatchDevicePath "/dev/input/event*"
        Driver "wacom"
        Option "Button2" "3"
```

btw, @danielquinn do you prefer a PR or just add the patch by yourself once is verified working for lower kernel version?

**Comment by vizv** [comment]
@joelacrisp I see a DBus event emitted once the top button of stylus is pressed, however bluetoothd crashes with a segfault after the press... Is this because I'm using KDE5? I have no idea how to get this working.

```
signal time=1463104449.039161 sender=:1.4 -> destination=(null destination) serial=2529 path=/; interface=org.kde.KDirNotify; member=FilesAdded
   string "bluetooth:/"
```

**Comment by joelacrisp** [comment]
I don't know sorry, I use AwesomeWM not KDE. Because it's awesome

**Comment by danielquinn** [comment]
Hi guys, I've been fiddling with this for weeks now and I just can't get my stylus to work (at all).  I've tried a number of variations on the config snippets above, but I think I've somehow bungled the kernel config and removed something important (though I have no idea what it might be)

Can someone give me a quick pointer on what's needed to get the stylus working at all?  Specifically:

* kernel config options
* packages installed
* x config
* anything else?

**Comment by joelacrisp** [comment]
For the touchscreen, it should "Just work", the blue button needs the Bluetooth drivers and userspace. Note: the Bluetooth button is weird. It uses a connect/disconnect as the signal for a press, not a connect...wait...button event...wait... eventual disconnect

**Comment by danielquinn** [comment]
My touch screen works great (even better since the upgrade to 4.6.1).  My problem is that the stylus doesn't register at all.  As in, when I run the pen across the screen, the mouse doesn't move.  I'm not even worrying about the buttons at this stage.

**Comment by cydergoth** [comment]
Battery in the stylus?

**Comment by danielquinn** [comment]
@cydergoth I don't think so.  The bluetooth connectivity still appears to work.  I can press-and-hold the purple button and it appears on my laptop as a bluetooth device and can even connect to it.  There's just no mouse movement on tap.

**Comment by joelacrisp** [comment]
Hmm, no idea then sorry
